mkdirp => fs.mkdirp; use built-in mkdir recursive

boneskull opened the pull request:
This PR:

1. Removes `mkdirp` dep from `@appium/support`
2. Creates the `fs.mkdirp` function in `@appium/support`
3. Deprecates the `mkdirp` module (re-exports `fs.mkdirp`)
4. Updates `appium` to use `fs.mkdirp`

boneskull committed:
feat(support): remove mkdirp

- `mkdirp` module is now deprecated and re-exports `fs.mkdirp`, which is new
- fix some types in `fs` module
- fix `walkDir` which did not actually resolve to `null` ever

boneskull committed:
chore(appium): use new mkdirp

- fix some types, mocks
- rename `HASHFILE_RELATIVE_PATH` to `PKG_HASHFILE_RELATIVE_PATH`; hopefully that's less opaque
- fixed broken junk in `commands-yml/parse.js` which will conflict with #16459 so idk

boneskull commented:
Current dependencies on/for this PR:
* 2.0
  * **PR #16443** <a href="https://app.graphite.dev/github/pr/appium/appium/16443" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #16459** <a href="https://app.graphite.dev/github/pr/appium/appium/16459" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #16482** <a href="https://app.graphite.dev/github/pr/appium/appium/16482" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).

mykola-mokhnach commented on packages/support/lib/fs.js:
```diff
@@ -16,14 +18,15 @@ import log from './logger';
     try {
-      await fs.access(path, _fs.R_OK);
+      // as of recent-ish node versions, `R_OK` moved into the `constants` property of `fs`
+      await fs.access(path, _fs.constants.R_OK);
```
shall we use `_fs.constants.R_OK ?? _fs.R_OK` ?

boneskull replied:
I guess the comment is rather ambiguous.  I'll fix that, but the changed one is correct

boneskull replied:
I just removed the comment.  `fs.R_OK` is deprecated, afaict.  guess I don't need to add a comment to explain code that no longer exists, huh

mykola-mokhnach commented on packages/support/lib/fs.js:
```diff
@@ -16,14 +18,15 @@ import log from './logger';
 
-const mkdirAsync = B.promisify(_fs.mkdir);
-const ncpAsync = B.promisify(ncp);
+const mkdirAsync = /** @type {(filepath: string, opts: _fs.MakeDirectoryOptions|undefined) => B<void>} */(B.promisify(_fs.mkdir));
```
can we use https://nodejs.org/api/fs.html#fspromisesmkdirpath-options, which returns a promise?

boneskull replied:
I was just keeping things the way they already are. that said: we can replace all of the promisified `fs` methods with native ones instead of doing it ourselves, unless there's a reason to wrap them in Bluebird promises...

mykola-mokhnach replied:
there was a reason for it before.:the promisified mkdir only appeared in node 10. Now, when we set node14 to be the minimum supported version it is totally safe to drop promisifying.

boneskull replied:
I think we're at Node.js v12 minimum now, so we can just use the builtins w/o promisifying.  I'll take care of that now.

boneskull replied:
@mykola-mokhnach I've made those changes.

mykola-mokhnach commented on packages/support/lib/fs.js:
```diff
@@ -45,15 +48,15 @@ const fs = {
    * @param {string} source - Source to copy
    * @param {string} destination - Destination to copy to
-   * @param {...any} args - Additional arguments to pass to `ncp`
+   * @param {ncp.Options} [opts] - Additional arguments to pass to `ncp`
```
would that be a breaking change? are these args used anywhere else in appium org?

boneskull replied:
no. `ncp` accepts four parameters, one of which is a callback which is promisified away.  `source`, `dest`, and `options`.  any further arguments are discarded, so it is not necessary to use a rest param here.

mykola-mokhnach commented on packages/support/lib/fs.js:
```diff
@@ -32,9 +35,9 @@ const fs = {
   rimraf: B.promisify(rimraf),
   rimrafSync: rimraf.sync.bind(rimraf),
-  async mkdir (...args) {
+  async mkdir (filepath, opts = {}) {
```
shall we check other org modules for possible usages of special args?

boneskull replied:
much like `ncp` above, `mkdir` only accepts three params (`dirname`, `opts`, and a `callback`).  so we only need the first two.

boneskull committed:
chore(support): use linkcode jsdoc tag since vscode understands it

boneskull committed:
chore(support): replace promisified fs functions with native fsPromises functions

(where applicable)

- typed and documented some stuff
- added `fs.openFile` which resolves w/a `fsPromises.FileHandle` object.  `fs.open` still resolves with a file descriptor.  this is, I think, the only explicit conflict between the builtin `fs` module and ours. we should consider deprecating or removing our `fs.open` and using the `FileHandle`-returning impl, because it's a little easier to use.  Once we do, we can also remove `fs.close()`, `fs.read()` and `fs.write()`, as those are redundant since they are methods of the `FileHandle` object returned from `fsPromises.open`.  A cursory check in the `appium` org suggests that `open` and its ilk are not used except in some test cases.
- assign properties directly to the `fs` object, as this is easier for the type system to understand (rather than iterating over property names and assigning properties in the loop)
- `which` actually returns a `Promise` now and does not need promisification
- added note about future changes to `rmdir`

jlipps reviewed (commented):
👍
